srml/rolling-window/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Rolling Window module
+//!
+//! ## Overview
+//!
+//! The Rolling Window Module is similar to `simple moving average` except
+//! that it just reports the number of occurrences in the window instead of
+//! calculating the average.
+//!
+//! It is mainly implemented to keep track of misbehaviors and only to take
+//! the last `sessions` of misbehaviors into account.
+//!
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(test)]
+mod mock;
+
+use srml_support::{
+	StorageMap, EnumerableStorageMap, decl_module, decl_storage,
+	traits::WindowLength
+};
+use parity_codec::Codec;
+use sr_primitives::traits::MaybeSerializeDebug;
+use srml_session::SessionIndex;
+
+/// Rolling Window trait
+pub trait Trait: system::Trait {
+	/// Kind to report with window length
+	type Kind: Copy + Clone + Codec + MaybeSerializeDebug + WindowLength<u32>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as RollingWindow {
+		/// Misbehavior reports
+		///
+		/// It stores every unique misbehavior of a kind
+		// TODO [#3149]: optimize how to shrink the window when sessions expire
+		MisbehaviorReports get(misbehavior_reports): linked_map T::Kind => Vec<SessionIndex>;
+
+		/// Bonding Uniqueness
+		///
+		/// Keeps track of uniquely reported misconducts in the entire bonding duration
+		/// which is currently unbounded (insert only)
+		///
+		/// Footprints need to be unique or stash accounts must be banned from joining
+		/// the validator set after been slashed
+		BondingUniqueness get(bonding_uniqueness): linked_map T::Hash => SessionIndex;
+	}
+}
+
+decl_module! {
+	/// Rolling Window module
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {}
+}
+
+/// Trait for getting the number of misbehaviors in the current window
+pub trait GetMisbehaviors<Kind> {
+	/// Get number of misbehavior's in the current window for a kind
+	fn get_misbehaviors(kind: Kind) -> u64;
+}
+
+impl<T: Trait> GetMisbehaviors<T::Kind> for Module<T> {
+	fn get_misbehaviors(kind: T::Kind) -> u64 {
+		<MisbehaviorReports<T>>::get(kind).len() as u64
+	}
+}
+
+/// Trait for reporting misbehavior's
+pub trait MisbehaviorReporter<Kind, Hash> {
+	/// Report misbehavior for a kind
+	fn report_misbehavior(kind: Kind, footprint: Hash, current_session: SessionIndex) -> Result<(), ()>;
+}
+
+impl<T: Trait> MisbehaviorReporter<T::Kind, T::Hash> for Module<T> {
+	fn report_misbehavior(
+		kind: T::Kind,
+		footprint: T::Hash,
+		current_session: SessionIndex,
+	) -> Result<(), ()> {
+		if <BondingUniqueness<T>>::exists(footprint) {
+			Err(())
+		} else {
+			<BondingUniqueness<T>>::insert(footprint, current_session);
+			<MisbehaviorReports<T>>::mutate(kind, |entry| entry.push(current_session));
+			Ok(())
+		}
+	}
+}
+
+impl<T: Trait> srml_session::OnSessionEnding<T::AccountId> for Module<T> {
```

Comment:
shouldn't this be a `impl SessionHandler`, for the function `on_new_session`?

Reply:
Ok, the only thing we need from this is the `session_number` in order to remove old misbehaviors and I don't see how `SessionHandler` could provide that?

I guess it could work by keeping a separate `SessionIndex` in this module?!

Reply:
you could just add it as a param of `fn on_new_session`

Reply:
Ok, I did it by modifying the function in the `traits`. I hope that is what you meant?